### PR TITLE
feature(gry): add MDQN algorithm

### PR DESCRIPTION
**Pull request by ruoyuGao**
## Description

Implementation of [Munchausen Reinforcement Learning](https://arxiv.org/abs/2007.14430)

### Reward
asterix

### Action gap
action_gap

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [x] pass all the tests

### CODE REVIEW

File: dizoo/atari/config/serial/Asterix/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from .asterix_mdqn_config import asterix_mdqn_config, asterix_mdqn_create_config
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/Asterix/__init__.py:1]: Asterix -> asterix

File: ding/rl_utils/tests/test_td.py [hunk 2]
```diff
@@ -585,3 +586,24 @@ def test_fn_td_lambda():
...
     assert tmp == reward.shape
+
+
+@pytest.mark.unittest
+def test_fn_m_q_1step_td_error():
+    batch_size = 128
+    action_dim = 9
+    q = torch.randn(batch_size, action_dim).requires_grad_(True)
+    target_q_current = torch.randn(batch_size, action_dim).requires_grad_(True)
+    target_q_next = torch.randn(batch_size, action_dim).requires_grad_(True)
```
Comment (PaParaZz1) [inline ding/rl_utils/tests/test_td.py:597]: target value should not requires grad

File: dizoo/atari/config/serial/Asterix/asterix_mdqn_config.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    exp_name='asterix_mdqn_seed0',
+    env=dict(
+        collector_env_num=8,
+        evaluator_env_num=8,
+        n_evaluator_episode=8,
+        stop_value=20000,
+        env_id='Asterix-v0',
+        #'ALE/SpaceInvaders-v5' is available. But special setting is needed after gym make.
+        frame_stack=4,
+        manager=dict(shared_memory=False, ),
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/Asterix/asterix_mdqn_config.py:14]: remove this, enable shm

File: dizoo/atari/config/serial/Asterix/asterix_mdqn_config.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        nstep=1,
+        discount_factor=0.99,
+        entropy_tau=0.03,
+        m_alpha=0.9,
+        learn=dict(
+            update_per_collect=4,
+            batch_size=32,
+            learning_rate=0.00005,
+            target_update_freq=8,
+            learner=dict(train_iterations=int(3e7), hook=dict(save_ckpt_after_iter=1000000, ))
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/Asterix/asterix_mdqn_config.py:33]: remove `train_iterations` in config, instead of indicating it in the argument of serial_pipeline (max_train_iter)

File: dizoo/atari/config/serial/Asterix/asterix_mdqn_config.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        ),
+    ),
+)
+asterix_mdqn_config = EasyDict(asterix_mdqn_config)
+main_config = asterix_mdqn_config
+asterix_mdqn_create_config = dict(
+    env=dict(
+        type='atari',
+        import_names=['dizoo.atari.envs.atari_env'],
+        reset_timeout=8,
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/Asterix/asterix_mdqn_config.py:54]: create config should only contain type and import_names

File: ding/policy/mdqn.py [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+
+from .base_policy import Policy
+from .common_utils import default_preprocess_learn
+
+
+@POLICY_REGISTRY.register('mdqn')
+class MDQNPolicy(Policy):
+    r"""
+    Overview:
+        Policy class of DQN algorithm, extended by Double DQN/Dueling DQN/PER/multi-step TD.
```
Comment (PaParaZz1) [inline ding/policy/mdqn.py:21]: modify info and add paper link

File: ding/policy/mdqn.py [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+from ding.model import model_wrap
+from ding.utils import POLICY_REGISTRY
+from ding.utils.data import default_collate, default_decollate
+
+from .base_policy import Policy
+from .common_utils import default_preprocess_learn
+
+
+@POLICY_REGISTRY.register('mdqn')
+class MDQNPolicy(Policy):
```
Comment (PaParaZz1) [inline ding/policy/mdqn.py:18]: maybe you can extend DQNPolicy to reuse some methods

File: ding/policy/mdqn.py [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+        # after update
+        # =============
+        self._target_model.update(self._learn_model.state_dict())
+        return {
+            'cur_lr': self._optimizer.defaults['lr'],
+            'total_loss': loss.item(),
+            'q_value': q_value.mean().item(),
+            'target_q_value': target_q_value.mean().item(),
+            'action_gap': action_gap.item(),
+            'clip_frac': clipfrac,
```
Comment (PaParaZz1) [inline ding/policy/mdqn.py:255]: should be a native scalar rather than tensor

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/dqn.py [hunk 9]
```diff
@@ -694,3 +695,237 @@ def _load_state_dict_learn(self, state_dict: Dict[str, Any]) -> None:
...
+class MDQNPolicy(DQNPolicy):
+    """
+    Overview:
+        Policy class of Munchausen DQN algorithm, extended by auxiliary objectives.
+        Paper link: https://arxiv.org/abs/2007.14430
+    Config:
+        == ==================== ======== ============== ======================================== =======================
+        ID Symbol               Type     Default Value  Description                              Other(Shape)
+        == ==================== ======== ============== ======================================== =======================
+        1  ``type``             str      dqn_stdim      | RL policy register name, refer to      | This arg is optional,
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:710]: modify config table

File: ding/policy/dqn.py [hunk 10]
```diff
@@ -694,3 +695,237 @@ def _load_state_dict_learn(self, state_dict: Dict[str, Any]) -> None:
...
+        priority=True,
+        # (bool) Whether use Importance Sampling Weight to correct biased update. If True, priority must be True.
+        priority_IS_weight=False,
+        # (float) Discount factor(gamma) for returns
+        discount_factor=0.97,
+        # (float) Entropy factor (tau) for Munchausen DQN
+        entropy_tau=0.03,
+        # (float) Discount factor (alpha) for Munchausen term
+        m_alpha=0.9,
+        nstep=1,
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:779]: add comments for each config field

File: ding/rl_utils/td.py [hunk 11]
```diff
@@ -40,6 +40,51 @@ def q_1step_td_error(
...
+    assert len(reward.shape) == 1, reward.shape
+    batch_range = torch.arange(act.shape[0])
+    if weight is None:
+        weight = torch.ones_like(reward)
+    q_s_a = q[batch_range, act]
+    # calculate muchausen addon
+    # replay_log_policy
+    target_v_s = target_q[batch_range].max(1)[0].unsqueeze(-1)
+    top2_q_s = target_q[batch_range].topk(2, dim=1, largest=True, sorted=True)[0]
+    action_gap = (top2_q_s[:, 0] - top2_q_s[:, 1]).mean()
```
Comment (PaParaZz1) [inline ding/rl_utils/td.py:65]: for something only for visualization, you can wrap it with torch.no_grad

File: dizoo/atari/config/serial/asterix/asterix_mdqn_config.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    exp_name='asterix_mdqn_seed0',
+    env=dict(
+        collector_env_num=1,
+        evaluator_env_num=8,
+        n_evaluator_episode=8,
+        stop_value=20000,
+        env_id='Asterix-v0',
+        #'ALE/SpaceInvaders-v5' is available. But special setting is needed after gym make.
+        frame_stack=4,
+        manager=dict(shared_memory=False, ),
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/asterix/asterix_mdqn_config.py:14]: enable shm

File: dizoo/atari/config/serial/asterix/asterix_mdqn_config.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    env_manager=dict(type='subprocess'),
+    policy=dict(type='mdqn'),
+)
+asterix_mdqn_create_config = EasyDict(asterix_mdqn_create_config)
+create_config = asterix_mdqn_create_config
+
+if __name__ == '__main__':
+    # or you can enter ding -m serial -c spaceinvaders_dqn_config.py -s 0
+    from ding.entry import serial_pipeline
+    serial_pipeline((main_config, create_config), seed=0)
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/asterix/asterix_mdqn_config.py:64]: add `max_env_step=int(1e7)`

File: ding/policy/dqn.py [hunk 14]
```diff
@@ -694,3 +695,237 @@ def _load_state_dict_learn(self, state_dict: Dict[str, Any]) -> None:
         self._target_model.load_state_dict(state_dict['target_model'])
         self._optimizer.load_state_dict(state_dict['optimizer'])
         self._aux_optimizer.load_state_dict(state_dict['aux_optimizer'])
+
+
+@POLICY_REGISTRY.register('mdqn')
+class MDQNPolicy(DQNPolicy):
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:701]: move to a file

File: ding/policy/dqn.py [hunk 15]
```diff
@@ -694,3 +695,237 @@ def _load_state_dict_learn(self, state_dict: Dict[str, Any]) -> None:
...
+        == ==================== ======== ============== ======================================== =======================
+    """
+    config = dict(
+        type='mdqn',
+        # (bool) Whether use cuda in policy
+        cuda=False,
+        # (bool) Whether learning policy is the same as collecting data policy(on-policy)
+        on_policy=False,
+        # (bool) Whether enable priority experience sample
+        priority=True,
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:770]: default to False

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/dqn.py (hunk 9 above)
Comment (ruoyuGao) [reply ding/policy/dqn.py:710]: modified

File: ding/policy/dqn.py (hunk 10 above)
Comment (ruoyuGao) [reply ding/policy/dqn.py:779]: added

File: ding/rl_utils/td.py (hunk 11 above)
Comment (ruoyuGao) [reply ding/rl_utils/td.py:65]: fixed

File: ding/rl_utils/tests/test_td.py [hunk 16]
```diff
@@ -585,3 +586,25 @@ def test_fn_td_lambda():
...
     assert tmp == reward.shape
+
+
+@pytest.mark.unittest
+def test_fn_m_q_1step_td_error():
+    batch_size = 128
+    action_dim = 9
+    q = torch.randn(batch_size, action_dim).requires_grad_(True)
+    target_q_current = torch.randn(batch_size, action_dim).requires_grad_(False)
+    target_q_next = torch.randn(batch_size, action_dim).requires_grad_(True)
```
[the remark this answers is not in the document]
Comment (ruoyuGao) [reply ding/rl_utils/tests/test_td.py:597]: fixed

File: ding/policy/dqn.py (hunk 10 above)
Comment (ruoyuGao) [reply ding/policy/dqn.py:779]: added

File: dizoo/atari/config/serial/asterix/asterix_mdqn_config.py (hunk 12 above)
Comment (ruoyuGao) [reply dizoo/atari/config/serial/asterix/asterix_mdqn_config.py:14]: changed

File: dizoo/atari/config/serial/asterix/asterix_mdqn_config.py (hunk 13 above)
Comment (ruoyuGao) [reply dizoo/atari/config/serial/asterix/asterix_mdqn_config.py:64]: added

File: ding/policy/dqn.py (hunk 15 above)
Comment (ruoyuGao) [reply ding/policy/dqn.py:770]: changed

File: ding/entry/serial_entry.py [hunk 17]
```diff
@@ -22,6 +22,7 @@ def serial_pipeline(
         model: Optional[torch.nn.Module] = None,
         max_train_iter: Optional[int] = int(1e10),
         max_env_step: Optional[int] = int(1e10),
+        is_dynamic_seed: Optional[bool] = None,
```
Comment (PaParaZz1) [inline ding/entry/serial_entry.py:25]: default to False

Comment (PaParaZz1) [inline ding/entry/serial_entry.py:25]: don't name variable like `is_xxx` or `use_xxx`, just `dynamic_seed`

File: ding/entry/tests/test_serial_entry.py [hunk 18]
```diff
@@ -68,6 +69,20 @@ def test_dqn():
...
+    config[0].policy.learn.update_per_collect = 1
+    config[0].exp_name = 'cartpole_mdqn_unittest'
+    try:
+        serial_pipeline(config, seed=0, max_train_iter=1, is_dynamic_seed=False)
+    except Exception:
+        assert False, "pipeline fail"
+    finally:
+        os.popen('rm -rf cartpole_mdqn_unittest')
+
+
```
Comment (PaParaZz1) [inline ding/entry/tests/test_serial_entry.py:85]: also add mdqn to algotest

File: ding/policy/mdqn.py [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+    def _init_learn(self) -> None:
+        """
+        Overview:
+            Learn mode init method. Called by ``self.__init__``, initialize the optimizer, algorithm arguments, main \
+            and target model.
+        """
+        self._priority = self._cfg.priority
+        self._priority_IS_weight = self._cfg.priority_IS_weight
+        # Optimizer
+        self._optimizer = Adam(self._model.parameters(), lr=self._cfg.learn.learning_rate, eps=0.0003125)
```
Comment (PaParaZz1) [inline ding/policy/mdqn.py:144]: add comments for why set eps

File: ding/rl_utils/td.py [hunk 20]
```diff
@@ -40,6 +40,58 @@ def q_1step_td_error(
...
+
+def m_q_1step_td_error(
+        data: namedtuple,
+        gamma: float,
+        tau: float,
+        alpha: float,
+        criterion: torch.nn.modules = nn.MSELoss(reduction='none')  # noqa
+) -> torch.Tensor:
+    q, target_q, next_q, act, reward, done, weight = data
+    lo = -1
```
Comment (PaParaZz1) [inline ding/rl_utils/td.py:54]: why name to `lo`

File: dizoo/atari/config/serial/asterix/asterix_mdqn_config.py [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+    exp_name='asterix_mdqn_seed0',
+    env=dict(
+        collector_env_num=8,
+        evaluator_env_num=8,
+        n_evaluator_episode=8,
+        stop_value=20000,
+        env_id='Asterix-v0',
+        #'ALE/SpaceInvaders-v5' is available. But special setting is needed after gym make.
+        frame_stack=4,
+        manager=dict(shared_memory=True, ),
```
Comment (PaParaZz1) [inline dizoo/atari/config/serial/asterix/asterix_mdqn_config.py:14]: remove this line, default config is True

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/rl_utils/td.py (hunk 20 above)
Comment (ruoyuGao) [reply ding/rl_utils/td.py:54]: because official implementation named lo here

File: ding/entry/serial_entry.py (hunk 17 above)
Comment (ruoyuGao) [reply ding/entry/serial_entry.py:25]: changed

Comment (ruoyuGao) [reply ding/entry/serial_entry.py:25]: changed

File: ding/entry/tests/test_serial_entry.py (hunk 18 above)
Comment (ruoyuGao) [reply ding/entry/tests/test_serial_entry.py:85]: added

File: ding/policy/mdqn.py (hunk 19 above)
Comment (ruoyuGao) [reply ding/policy/mdqn.py:144]: addded

File: ding/rl_utils/td.py (hunk 20 above)
Comment (ruoyuGao) [reply ding/rl_utils/td.py:54]: changed

File: dizoo/atari/config/serial/asterix/asterix_mdqn_config.py (hunk 21 above)
Comment (ruoyuGao) [reply dizoo/atari/config/serial/asterix/asterix_mdqn_config.py:14]: removed

**Review by PaParaZz1 — APPROVED** [review]

**Review by PaParaZz1 — APPROVED** [review]
